nemo/collections/asr/modules/flashlight_decoder.py

```diff
@@ -0,0 +1,253 @@
+        }
+
+    @property
+    def output_types(self):
+        """Returns definitions of module output ports.
+        """
+        return {"hypos": NeuralType(('B'), PredictionsType())}
+    '''
+    
+    class TokensWrapper:
+        def __init__(self, vocabulary: List[str], tokenizer: TokenizerSpec):
+            self.vocabulary = vocabulary
+            self.tokenizer = tokenizer
+            
+            if tokenizer is None:
+                self.reverse_map = {vocabulary[i]:i for i in range(len(vocabulary))}
+        
+        @property
+        def blank(self):
+            return len(self.vocabulary)
+        
+        @property
+        def unk_id(self):
+            if (self.tokenizer is not None) and hasattr(self.tokenizer, 'unk_id'):
+                return self.tokenizer.unk_id
+            
+            if '<unk>' in self.vocabulary:
+                return self.token_to_id('<unk>')
+            else:
+                return -1
+        
+        @property
+        def vocab(self):
+            return self.vocabulary
+        
+        @property
+        def vocab_size(self):
+            return len(self.vocabulary) + 1
+        
+        def token_to_id(self, token: str):
+            if token == self.blank:
+                return -1
+            
+            if self.tokenizer is not None:
+                return self.tokenizer.token_to_id(token)
+            else:
+                return self.reverse_map[token]
+    
+    
+    def __init__(
+        self,
+        lm_path: str,
+        vocabulary: List[str],
+        tokenizer: Optional[TokenizerSpec] = None,
+        lexicon_path: Optional[str] = None,
+        nbest: int = 1,
+        beam_size:int = 32,
+        beam_size_token: int = 32,
+        beam_threshold: float = 25.0,
+        lm_weight: float = 2.0,
+        word_score: float = -1.0,
+        unk_weight: float = -math.inf,
+        sil_weight: float = 0.0,
+        unit_lm: bool = False,
+    ):
+
+        try:
+            from flashlight.lib.text.dictionary import create_word_dict, load_words
+            from flashlight.lib.sequence.criterion import get_data_ptr_as_bytes
+            from flashlight.lib.text.decoder import (
+                CriterionType,
+                LexiconDecoderOptions,
+                KenLM,
+                LM,
+                LMState,
+                SmearingMode,
+                Trie,
+                LexiconDecoder,
+            )
+        except ModuleNotFoundError:
+            raise ModuleNotFoundError(
+                "FlashLightKenLMBeamSearchDecoder requires the installation of flashlight python bindings "
+                "from https://github.com/flashlight/text. Please follow the build instructions there."
+            )
+
+        super().__init__()
+        
+        self.criterion_type = CriterionType.CTC
+        self.nbest = nbest
+        self.tokenizer = self.TokensWrapper(vocabulary, tokenizer)

```

Comment:
Maybe more natural to call it `self.tokenizer_wrapper` to avoid the notion that its the actual tokenizer

Reply:
Done